### PR TITLE
[Enhancement] improve error message handling of JNI connector and support char type of paimon connector

### LINKED ISSUE #27992: [paimon catalog] error from paimon connector: Failed to open the off-heap table scanner.: file =
### Steps to reproduce the behavior (Required)
1. create paimon table and insert data:
```
create table paimon_hdfs_orc_none(
 col_tinyint tinyint comment "column tinyint"
,col_smallint smallint comment "column smallint"
,col_int int comment "column int"
,col_integer integer comment "column integer"
,col_bigint bigint comment "column bigint"
,col_float float comment "column float"
,col_double double comment "column double"
,col_double_precision double precision comment "column double precision"
,col_decimal decimal(10,2) comment "column decimal"
,col_timestamp timestamp comment "column timestamp"
,col_date date comment "column date"
,col_string string comment "column string"
,col_varchar varchar(100) comment "column varchar"
,col_binary binary comment "column binary"
,col_char char(100) comment "column char"
,col_boolean boolean comment "column boolean"
,col_array array<int> comment "column array"
)
STORED BY 'org.apache.paimon.hive.PaimonStorageHandler'
TBLPROPERTIES (
    'primary-key' = 'col_tinyint,col_date',
    'partition'='col_tinyint',
    'bucket' = '2',
    'file.format' = 'orc',
    'file.compression' = 'none'
);
insert into paimon_hdfs_orc_none select 1,1,1,1,1,1.001,10.1,110.1,1110.1,'2021-10-30 12:10:23','2021-10-30','hello world','hi hello','1110001010101011001001','nihao',true,array(1,10,100);
```

2. create catalog in sr;
3. select paimon table through paimon catalog error:
```
mysql> select count(*) from paimon_hdfs_orc_none;
ERROR 1064 (HY000): Failed to open the off-heap table scanner.: file =

mysql> select count(col_tinyint) from paimon_hdfs_orc_none;
ERROR 1064 (HY000): Failed to open the off-heap table scanner.: file =

mysql> select col_tinyint from paimon_hdfs_orc_none;
ERROR 1064 (HY000): Failed to open the off-heap table scanner.: file =
``` 

### Expected behavior (Required)
success

### Real behavior (Required)
fail

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
branch-3.1-a0d31c9
```

### PR DESCRIPTION
**Pull request by miomiocat**
Fixes #27992

- fail fast for unsupported types of paimon table
- output error massage of JNI to client terminal
- support char type of paimon connector
- some minor refactor

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 14 (85.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|12|14|85.71%|[462, 468]|

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -523,6 +523,10 @@ Status HiveDataSource::_init_scanner(RuntimeState* state) {
     RETURN_IF_ERROR(scanner->init(state, scanner_params));
     Status st = scanner->open(state);
     if (!st.ok()) {
+        if (typeid(*scanner) == typeid(JniScanner)) {
```
Comment (dirtysalt) [inline be/src/connector/hive_connector.cpp:526]: how about a virtual function instead of this way?

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|11|11|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|1|2|50.00%|[397]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|2|2|100.00%|[]|

**Review by dirtysalt — APPROVED** [review]

**Review by caneGuy — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**